cpp/src/lance/arrow/updater.cc

```diff
@@ -0,0 +1,209 @@
+
+  ::arrow::Status Update(const std::shared_ptr<::arrow::Array>& arr);
+
+  ::arrow::Result<std::shared_ptr<LanceDataset>> Finish();
+
+ private:
+  auto data_dir() const { return dataset_->impl_->data_dir(); }
+
+  const auto& fs() const { return dataset_->impl_->fs; }
+
+  /// Prepare to read the next fragment.
+  ::arrow::Status NextFragment();
+
+  std::shared_ptr<LanceDataset> dataset_;
+  /// The schema of the newly generated dataset.
+  std::shared_ptr<lance::format::Schema> full_schema_;
+  /// The schema of the column to updated/added
+  std::shared_ptr<lance::format::Schema> column_schema_;
+  /// A copy of fragments.
+  ::arrow::dataset::FragmentVector fragments_;
+
+  // Used to store the updated fragments.
+  std::vector<std::shared_ptr<format::DataFragment>> data_fragments_;
+
+  // Track runtime information.
+
+  ::arrow::dataset::FragmentVector::iterator fragment_it_;
+  std::shared_ptr<::arrow::RecordBatch> last_batch_;
+  std::unique_ptr<lance::io::FileWriter> writer_;
+  ::arrow::RecordBatchGenerator batch_generator_;
+};
+
+::arrow::Status Updater::Impl::NextFragment() {
+  assert(!writer_);
+  auto& fragment = *fragment_it_;
+  assert(fragment->type_name() == "lance");
+  auto lance_fragment = std::dynamic_pointer_cast<LanceFragment>(fragment);
+  assert(lance_fragment);
+
+  std::string file_path = fs::path(data_dir()) / fmt::format("{}.lance", GetUUIDString());
+  ARROW_ASSIGN_OR_RAISE(auto output, fs()->OpenOutputStream(file_path));
+  auto write_options = LanceFileFormat().DefaultWriteOptions();
+  writer_ =
+      std::make_unique<io::FileWriter>(column_schema_, std::move(write_options), std::move(output));
+
+  ARROW_ASSIGN_OR_RAISE(auto scan_builder, dataset_->NewScan());
+  //  ARROW_RETURN_NOT_OK(scan_builder->BatchSize(std::numeric_limits<int64_t>::max()));
+  ARROW_ASSIGN_OR_RAISE(auto scanner, scan_builder->Finish());
+  ARROW_ASSIGN_OR_RAISE(batch_generator_, (*fragment_it_)->ScanBatchesAsync(scanner->options()));
+
+  // Track the new data files
+  std::vector<format::DataFile> data_files(lance_fragment->data_fragment()->data_files());
+  data_files.emplace_back(::lance::format::DataFile({file_path, column_schema_->GetFieldIds()}));
+  auto new_fragment = std::make_shared<format::DataFragment>(data_files);
+  data_fragments_.emplace_back(std::move(new_fragment));
+
+  return ::arrow::Status::OK();
+}
+
+::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> Updater::Impl::Next() {
+  if (fragment_it_ == fragments_.end()) {
+    return nullptr;
+  }
+  if (last_batch_) {
+    return ::arrow::Status::IOError("Have not consumed/committed last batch");
+  }
+  while (!last_batch_ && fragment_it_ != fragments_.end()) {
+    if (!writer_) {
+      ARROW_RETURN_NOT_OK(NextFragment());
+    }
+
+    auto fut = batch_generator_();
+    ARROW_ASSIGN_OR_RAISE(last_batch_, fut.result());
+    if (!last_batch_) {
+      ARROW_RETURN_NOT_OK(writer_->Finish().result());
```

Comment:
i thought this was reading batches from original dataset. why is a writer involved here?

Reply:
An updater is basically an iterator to hold a "Batch Generator" and "Batch Writer", the two of which will be opened / closed together to make data aligned between "Data Files".